src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs

```diff
@@ -0,0 +1,2408 @@
+                goto Throw;
+            }
+            else
+            {
+                if (!literal.StartsWith(span))
+                {
+                    _bytePositionInLine += FindMismatch(span, literal);
+                    goto Throw;
+                }
+
+                ReadOnlySpan<byte> leftToMatch = literal.Slice(span.Length);
+
+                SequencePosition startPosition = _currentPosition;
+                int startConsumed = _consumed;
+                int alreadyMatched = literal.Length - leftToMatch.Length;
+                while (true)
+                {
+                    _totalConsumed += alreadyMatched;
+                    _bytePositionInLine += alreadyMatched;
+                    if (!GetNextSpan())
+                    {
+                        _totalConsumed = prevTotalConsumed;
+                        consumed = default;
+                        if (IsLastSpan)
+                        {
+                            goto Throw;
+                        }
+                        return false;
+                    }
+
+                    span = _buffer;
+
+                    if (span.StartsWith(leftToMatch))
+                    {
+                        HasValueSequence = true;
+                        SequencePosition start = new SequencePosition(startPosition.GetObject(), startPosition.GetInteger() + startConsumed);
+                        SequencePosition end = new SequencePosition(_currentPosition.GetObject(), _currentPosition.GetInteger() + leftToMatch.Length);
+                        ValueSequence = _sequence.Slice(start, end);
+                        consumed = leftToMatch.Length;
+                        return true;
+                    }
+
+                    if (!leftToMatch.StartsWith(span))
+                    {
+                        _bytePositionInLine += FindMismatch(span, leftToMatch);
+                        goto Throw;
+                    }
+
+                    leftToMatch = leftToMatch.Slice(span.Length);
+                    alreadyMatched = span.Length;
+                }
+            }
+        Throw:
+            _totalConsumed = prevTotalConsumed;
+            consumed = default;
+            ThrowInvalidLiteralMultiSegment(span);
+            return false;
+        }
+
+        private int FindMismatch(ReadOnlySpan<byte> span, ReadOnlySpan<byte> literal)
+        {
+            Debug.Assert(span.Length > 0 && span[0] == literal[0]);
+
+            int indexOfFirstMismatch = 0;
+
+            for (int i = 1; i < literal.Length; i++)
+            {
+                if (span.Length > i)
+                {
+                    if (span[i] != literal[i])
+                    {
+                        indexOfFirstMismatch = i;
+                        break;
+                    }
+                }
+                else
+                {
+                    indexOfFirstMismatch = i;
+                    break;
+                }
+            }
```

Comment:
Rather than branching on every iteration, would it be better to iterate to Min(span.Length, literal.Length), and then if there's no mismatch up until that point and one is longer than the other, that's the position of the first mismatch?

Reply:
This is on the exception code path, so the performance doesn't matter too much here (especially since the literal is at most length 5). But will fix.